src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+        r.lifetime = dns_timeout
+        r.timeout = r.lifetime / 3
+
+        try:
+
+            answers = r.query(host, 'SRV')
+            srv_records = sorted(
+                answers,
+                key=lambda a: (int(a.priority), int(a.weight))
+            )
+
+        except Exception as e:
+            srv_records = []
+
+        return srv_records
+
+    def port_is_listening(self, host, port, timeout=1):
+        ret = False
+
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        if timeout:
+            s.settimeout(timeout)
+
+        try:
+            s.connect((host, port))
+            ret = True
+
+        except Exception as e:
+            s.close()
+            raise CallError(e)
+
+        s.close()
+        return ret
+
+    def _get_servers(self, srv_prefix):
+        """
+        We will first try fo find servers based on our AD site. If we don't find
+        a server in our site, then we populate list for whole domain. Ticket #27584
+        Domain Controllers, Forest Global Catalog Servers, and Kerberos Domain Controllers
+        need the site information placed before the 'msdcs' component of the host entry.t
+        """
+        servers = []
+        if not self.ad['domainname']:
+            return servers 
+
+        if self.ad['site'] and self.ad['site'] != 'Default-First-Site-Name':
+            if 'msdcs' in srv_prefix.value: 
+                parts = srv_prefix.value.split('.')
+                srv = '.'.join([parts[0],parts[1]])
+                msdcs = '.'.join([parts[2], parts[3]])
+                host = f"{srv}.{self.ad['site']}._sites.{msdcs}.{self.ad['domainname']}"
+            else:
+                host = f"{srv_prefix.value}{self.ad['site']}._sites.{self.ad['domainname']}"
+        else:
+            host = f"{srv_prefix.value}{self.ad['domainname']}"
+
+        servers = self._get_SRV_records(host, self.ad['dns_timeout'])
+
+        if not servers and self.ad['site']:
+            host = f"{srv_prefix.value}{self.ad['domainname']}"
+            dcs = self._get_SRV_records(host, self.ad['dns_timeout'])
+
+        if SSL(self.ad['ssl']) == SSL.USESSL:
+            for server in servers:
+                if server.port == 389:
+                    server.port = 636
+
+        return servers
+        
+    def get_n_working_servers(self, srv=SRV['DOMAINCONTROLLER'], number=1):
+        servers = self._get_servers(srv)
+        found_servers = []
+        for server in servers:
+            if len(found_servers) == number:
+                continue
+
+            host = server.target.to_text(True)
+            port = int(server.port)
+            if self.port_is_listening(host, port, timeout=1):
+                server_info = {'host': host, 'port': port}
+                found_servers.append(server_info)
+
+        if self.ad['verbose_logging']:
+            self.logger.debug(f'Request for [{number}] of server type [{srv.name}] returned: {found_servers}')
```

Comment:
For future reference, we should use % formatting in logger because it will be useful to aggregate the same kind of message but with different args.